cdap-api/src/main/java/co/cask/cdap/api/ProgramSpecificationBase.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api;
+
+import co.cask.cdap.api.plugin.Plugin;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Base class for {@link ProgramSpecification} which contains common information for various program specifications.
+ */
+public class ProgramSpecificationBase implements ProgramSpecification {
```

Comment:
So, why this class is not abstract then? It seems only for sub-classing. Also, rename to `AbstractProgramSpecificaton` if it is abstract.

Reply:
It is not abstract as in it does not exposes any new method which is abstract or omits to provide the implementation of any methods exposed by ProgramSpecification making it abstract. The initial idea to make it abstract was to disallow any instance creation of this class but was changed as it does not have any abstract method but it looks like officially a class does not have to contain an abstract method to be called Abstract.

```
An abstract class is a class that is declared abstract—it may or may not include abstract methods. Abstract classes cannot be instantiated, but they can be subclassed.
```
https://docs.oracle.com/javase/tutorial/java/IandI/abstract.html

Reply:
The question is, do you intent someone new it directly? If not, make it abstract.

Reply:
Yes, it is not intended that someone will new it directly. I made it Abstract in my last commit. Please have a look.